libnymea-core/scriptengine/scriptthing.cpp

```diff
@@ -0,0 +1,195 @@
+    if (actionType.id().isNull()) {
+        qCWarning(dcScriptEngine()) << "Thing" << thing->name() << "does not have action" << actionName;
+        return;
+    }
+
+    Action action(actionType.id(), thing->id(), Action::TriggeredByScript);
+    ParamList paramList;
+    foreach (const QString &paramNameOrId, params.keys()) {
+        ParamType paramType;
+        if (!ParamTypeId(paramNameOrId).isNull()) {
+            paramType = actionType.paramTypes().findById(ParamTypeId(paramNameOrId));
+        } else {
+            paramType = actionType.paramTypes().findByName(paramNameOrId);
+        }
+        if (paramType.id().isNull()) {
+            qCWarning(dcScriptEngine()) << "Invalid param id or name";
+            continue;
+        }
+        paramList << Param(paramType.id(), params.value(paramNameOrId));
+    }
+    action.setParams(paramList);
+    qCDebug(dcScriptEngine()) << "Executing action:" << action.thingId() << action.actionTypeId() << action.params();
+    m_thingManager->executeAction(action);
+
+}
+
+void ScriptThing::init(ThingManager *thingManager)
+{
+    m_thingManager = thingManager;
+    connect(m_thingManager, &ThingManager::thingAdded, this, [this](Thing *newThing){
+        if (newThing->id() == ThingId(m_thingId)) {
+            qCDebug(dcScriptEngine()) << "Thing" << newThing->name() << "appeared in system";
+            connectToThing();
+        }
+    });
+    connect(m_thingManager, &ThingManager::thingStateChanged, this, [=](Thing *thing, const StateTypeId &stateTypeId, const QVariant &value, const QVariant &minValue, const QVariant &maxValue){
+        Q_UNUSED(minValue)
+        Q_UNUSED(maxValue)
+        if (m_thingId != thing->id()) {
+            return;
+        }
+        emit stateValueChanged(thing->thingClass().getStateType(stateTypeId).name(), value);
+    });
+    connect(m_thingManager, &ThingManager::eventTriggered, this, [=](const Event &event){
+        if (ThingId(m_thingId) != event.thingId()) {
+            return;
+        }
+
+        Thing *thing = m_thingManager->findConfiguredThing(event.thingId());
+        QVariantMap params;
+        foreach (const Param &param, event.params()) {
+            params.insert(param.paramTypeId().toString().remove(QRegExp("[{}]")), param.value().toByteArray());
+            QString paramName = thing->thingClass().eventTypes().findById(event.eventTypeId()).paramTypes().findById(param.paramTypeId()).name();
+            params.insert(paramName, param.value().toByteArray());
+        }
+
+        // Note: Explicitly convert the params to a Json document because auto-casting from QVariantMap to the JS engine might drop some values.
+        emit eventTriggered(thing->thingClass().eventTypes().findById(event.eventTypeId()).name(), QJsonDocument::fromVariant(params).toVariant().toMap());
+    });
+}
+
+void ScriptThing::connectToThing()
+{
+    disconnect(m_nameConnection);
+
+    Thing *thing = m_thingManager->findConfiguredThing(ThingId(m_thingId));
```

Comment:
`m_thingId` seems to be already from type `ThingId`

Reply:
done